### DISCUSSION TITLE
How to use MsQuic on Xamarin and Linux .Net project?

### DISCUSSION THREAD

**Discussion by trudyhood** [microsoft/msquic#3579] [Q&A]
I need to know how to add MsQuic to a Xamarin and Linux .NET project, Where can I find a sample?

**Comment by nibanks** [comment]
Hi @trudyhood. Please see https://github.com/microsoft/quicspeed which is an old project we put together that does much of what you need. You should be able to take what you need from it.

**Comment by trudyhood** [comment]
Thank you, I looked at it but couldn't find anything about using System.Net.Quic in the Xamarin project.
It may be beyond my expertise.
I could make System.Net.Quic work on Linux, as it just needed an additional module to be installed on the OS, but I still have problems on Android. It doesn't detect ystem.Net.Quic and I couldn't find any nuget to install.

**Reply by nibanks** [reply]
Yeah, it doesn't use it, but instead calls MsQuic directly. The .NET folks (@wfurt, @ManickaP, @CarnaViire) might be able to help provide guidance for System.Net.Quic (which is still in "preview" support, but mostly stable).

**Reply by ManickaP** [reply]
System.Net.Quic is PNSE lib on Android, i.e. the implementation of all public APIs just throws `PlatformNotSupportedException`, there are no method bodies. That's the reason why it doesn't light up even with MsQuic present.

As far as I know, there are no plans to support System.Net.Quic on Android (or any other mobile platform). Currently we support: Windows 11+ (due to SChannel support of TLS 1.3), Linux ("should" work everywhere where MsQuic is) and tentatively MacOS. Side note, our distro coverage in 7.0 was limited, so you might encounter some issues; we improved that in 8.0 a lot.

You theoretically could try to rebuild System.Net.Quic with Linux implementation for Android and try to use it, but "here be dragons". None of us ever tried that.

**Reply by trudyhood** [reply]
So pity the majority of our users use Android. I hope to see it sooner on Android.

**Comment by wfurt** [comment]
I think it is mostly size and dependency problem. AFAIK There is Quic support for Android but it is not part of the base OS. We _could_ drag OpenSSL + MsQuic but that takes space and responsibility for patching. 

 cc: @steveisok @akoeplinger

**Reply by akoeplinger** [reply]
Yeah shipping OpenSSL is not an option for us so unless there's some Android  API that exposes what we need our hands are tied.
